### PR TITLE
fix(dashpay): DPNS normalization, contact resolution, key type, and privateData fixes

### LINKED ISSUE #687: impossible to Add Contact by Username
on Add Contact screen we can add concat by Username or Identity ID, but when we set existing Username it couldn't be found

Image
Image

cc: @thepastaclaw  @lklimek

### PR DESCRIPTION
**Pull request by lklimek**
## Summary

- **Fix DPNS profile search** — used `to_lowercase()` but DPNS stores `normalizedLabel` using homograph-safe conversion (`o`→`0`, `i`/`l`→`1`). Any name containing these characters would never match.
- **Fix "Add Contact" username resolution** — missing `normalizedParentDomainName == "dash"` compound filter in DPNS query, plus wrong normalization. Also used `owner_id()` instead of `records.identity` (wrong after name transfers).
- **Fix contact request rejection (key type)** — `create_or_update_contact_info()` only accepted `ECDSA_SECP256K1` signing keys. Many Platform identities use BLS keys. Verified against DIP-0015: no key type restriction for contactInfo documents.
- **Fix contact request rejection (privateData too small)** — DashPay contract requires `privateData` 48-2048 bytes. Reject with empty nickname/note produced only 32 bytes. Padded with 0x00 sentinel + random bytes.
- **Case-insensitive `.dash` suffix handling** — `strip_suffix(".dash")` and `ends_with(".dash")` were case-sensitive. `"Alice.DASH"` would fail.
- **Centralize DPNS normalization** — new `model::dpns` helper with `normalize_dpns_label()`, `strip_dash_suffix()`, `has_dash_suffix()`, `validate_dpns_input()`. All DPNS-by-name lookup sites use the shared helpers. 9 unit tests.
- **Centralize username input validation** — UI and backend both call `model::dpns::validate_dpns_input()` instead of implementing their own `.dash` suffix checks. Fixes case-sensitive check in Add Contact screen.
- **Use cached DPNS contract** — `resolve_username_to_identity()` had a hardcoded Base58 contract ID and fetched the contract from network on every call. Now uses the cached `app_context.dpns_contract`.
- **Early username format validation** — inputs like `"alice.foo"` (non-`.dash` domain) now return `InvalidUsername` immediately instead of making a wasted DPNS query.
- **Contact info size validation** — calls `validate_contact_info_field_sizes()` after encryption to catch oversized data before broadcasting.
- **Reject self-contact requests** — Platform rejects with code 40500 ("Identity must not be equal to owner id"). Now caught before broadcasting with a clear message.

## Fixes

Fixes #687
Supersedes #723

## Changes by File

| File | Change |
|---|---|
| `src/model/dpns.rs` | **New**: DPNS normalization + validation helpers with 9 unit tests |
| `src/model/mod.rs` | Register `dpns` module |
| `src/backend_task/dashpay/profile.rs` | Use `normalize_dpns_label()` for search |
| `src/backend_task/dashpay/contact_requests.rs` | Use `normalize_dpns_label()` + `has_dash_suffix()` + `validate_dpns_input()`; add `normalizedParentDomainName` filter; use `records.identity` instead of `owner_id()`; use cached DPNS contract; reject self-contact |
| `src/backend_task/identity/load_identity_by_dpns_name.rs` | Use `normalize_dpns_label()` |
| `src/backend_task/dashpay/contact_info.rs` | Accept all key types for signing; pad privateData to 48-byte minimum (0x00 sentinel + random); validate encrypted field sizes before broadcast |
| `src/backend_task/dashpay/errors.rs` | Add `CannotContactSelf` and `ContactInfoValidationFailed` variants |
| `src/ui/dashpay/add_contact_screen.rs` | Use `validate_dpns_input()` instead of case-sensitive `ends_with(".dash")` |
| `CLAUDE.md` | Add validation placement convention |

### DPNS Normalization
DPNS normalizes labels with `convert_to_homograph_safe_chars()` (o→0, i/l→1 + lowercase). All lookup sites were using plain `to_lowercase()`.

### Missing DPNS Index Filter
`resolve_username_to_identity()` queried `normalizedLabel` without the required `normalizedParentDomainName == "dash"` compound filter — the DPNS index needs both fields.

### Wrong Identity Reference
`resolve_username_to_identity()` used `document.owner_id()` to get the identity from a DPNS document. After a name transfer, `owner_id()` still points to the original registrant. `records.identity` is the authoritative reference.

### Hardcoded DPNS Contract
`resolve_username_to_identity()` hardcoded the DPNS contract ID and fetched it from network on every call, while `AppContext` already caches it at startup.

### Key Type Restriction
`create_or_update_contact_info()` restricted signing keys to ECDSA_SECP256K1. The DashPay contract (DIP-0015) has no key type constraint for contactInfo. Platform accepts all key types.

### privateData Minimum Size
When rejecting a contact (no nickname, no note), serialized privateData was 8 bytes → 32 bytes after AES-CBC. Contract requires ≥48 bytes.

### Case-Sensitive UI Validation
`add_contact_screen.rs` used `ends_with(".dash")` (case-sensitive) to validate usernames. Inputs like `olivia22.DASH` were rejected before reaching the backend's correct case-insensitive check.

## Test plan

- [x] `cargo clippy --all-features --all-targets -- -D warnings` clean
- [x] 518 tests pass (433 unit + 10 e2e + 72 kittest + 3 doc)
- [x] 9 unit tests for DPNS normalization + validation helpers
- [ ] Manual: Scenario 1 — DPNS name resolution (mixed case, whitespace, homograph chars, non-.dash domains)
- [ ] Manual: Scenario 2 — Contact request send/accept/reject flow
- [ ] Manual: Scenario 3 — BLS-only identity contact operations

<sub>🤖 Co-authored by [Claudius the Magnificent](https://github.com/lklimek/claudius) AI Agent</sub>

### CODE REVIEW

**Comment by lklimek** [comment]
refs #687

**Comment by thepastaclaw** [comment]
✅ Review complete (commit cd90c212)

File: src/model/dpns.rs [hunk 1]
```diff
@@ -0,0 +1,93 @@
+//! DPNS name normalization helpers.
```
Comment (thepastaclaw) [inline src/model/dpns.rs:1]: **🔴 Blocking: UTF-8 panic in strip_dash_suffix and has_dash_suffix (CONVERGENT — NOT FIXED)**

Both functions slice strings at byte offsets (input.len() - DASH_SUFFIX.len()) without checking is_char_boundary(). For inputs like "tëstab" (7 bytes), the split point (2) falls inside the 2-byte ë character → panic. Fix commit cb4652e8 claims to address this but only extracted a constant; the byte-offset slicing is identical. DPNS labels are ASCII-only on-chain, but these functions run on raw user input BEFORE platform validation, so any multi-byte character in search/contact fields triggers an unrecoverable crash.

<sub>source: ['claude']</sub>

<details>
<summary>🤖 Fix this with AI agents</summary>

```
These findings are from an automated code review. Verify each finding against the current code and only fix it if needed.

In `src/model/dpns.rs`:
- [BLOCKING] line 1: UTF-8 panic in strip_dash_suffix and has_dash_suffix (CONVERGENT — NOT FIXED)
  Both functions slice strings at byte offsets (input.len() - DASH_SUFFIX.len()) without checking is_char_boundary(). For inputs like "tëstab" (7 bytes), the split point (2) falls inside the 2-byte ë character → panic. Fix commit cb4652e8 claims to address this but only extracted a constant; the byte-offset slicing is identical. DPNS labels are ASCII-only on-chain, but these functions run on raw user input BEFORE platform validation, so any multi-byte character in search/contact fields triggers an unrecoverable crash.
```
</details>

Comment (thepastaclaw) [inline src/model/dpns.rs:1]: **💬 Nitpick: API asymmetry: has_dash_suffix trims whitespace, strip_dash_suffix does not**

has_dash_suffix() internally trims its input (line 42) while strip_dash_suffix() does not. This is mitigated in the main code path (normalize_dpns_label trims first), but direct callers of strip_dash_suffix could get unexpected results with whitespace-padded input.

<sub>source: ['claude']</sub>

**Review by thepastaclaw — CHANGES_REQUESTED** [review]
## Code Review

PR #810 introduces DPNS normalization helpers, DashPay contact info encryption, and key type flexibility. Six findings verified against source; three are blocking. The convergent UTF-8 panic finding is confirmed STILL PRESENT — the fix commit cb4652e8 only replaced magic numbers with a constant but did not add `is_char_boundary` guards. Two additional medium-severity issues: rootEncryptionKeyIndex stores the wrong key ID, and a hardcoded DPNS contract ID breaks non-mainnet usage.

*Reviewed commit: a9c07626*

🔴 1 blocking | 🟡 2 suggestion(s) | 💬 3 nitpick(s)

### 4 additional findings

<details>
<summary>🟡 <b>suggestion:</b> rootEncryptionKeyIndex written from authentication signing key, not encryption root key</summary>

`src/backend_task/dashpay/contact_info.rs (line 1)`

Line 435 writes `signing_key.id()` into `rootEncryptionKeyIndex`. The signing_key is an AUTHENTICATION-purpose key (lines 413-424), but rootEncryptionKeyIndex per DIP-15 should reference the identity key used as the root of the encryption key derivation tree. The read path at line 353 ignores this value (`_root_idx`), so self-decryption works, but other DashPay clients (e.g., mobile wallets) that rely on this field to find the correct encryption key will fail to decrypt contactInfo.

</details>
<details>
<summary>🟡 <b>suggestion:</b> Hardcoded DPNS contract ID breaks non-mainnet contact requests</summary>

`src/backend_task/dashpay/contact_requests.rs (line 1)`

resolve_username_to_identity() hardcodes the mainnet DPNS contract ID (GWRSAVFMjXx8HpQFaNJMqBV7MBgMK4br5UESsB4S31Ec) and re-fetches it from the network, while every other DPNS lookup uses app_context.dpns_contract (which is network-aware). On testnet/devnet, this hardcoded ID won't match, causing contact request username resolution to fail.

</details>
<details>
<summary>💬 <b>nitpick:</b> Sentinel-based padding in ContactInfoPrivateData lacks corresponding deserializer</summary>

`src/backend_task/dashpay/contact_info.rs (line 1)`

The serializer pushes a 0x00 sentinel byte before random padding (lines 88-94), but ContactInfoPrivateData has no deserialize() method. decrypt_private_data() exists (line 281) and is currently #[allow(dead_code)]. The sentinel design is correct for the current format (0x00 after the accepted_accounts section unambiguously marks padding), but without a deserializer, the correctness of the sentinel protocol is unverifiable.

</details>
<details>
<summary>💬 <b>nitpick:</b> Devnet/regtest SPV silently proceeds with no peers when core_host is unconfigured</summary>

`src/spv/manager.rs (line 1)`

With NetworkConfig fields now Optional, a devnet config without core_host parses successfully but primary_peer_socket() returns None (line 1398 does `config.core_host.as_deref()?`). The devnet branch at line 1346-1349 silently skips peer addition, leaving SPV with no peers and no DNS seeds — it will never sync. Previously, missing core_host would fail config parsing entirely. This is a design trade-off (app starts vs config error), not a strict regression.

</details>

<details>
<summary>🤖 Prompt for all review comments with AI agents</summary>

```
These findings are from an automated code review. Verify each finding against the current code and only fix it if needed.

In `src/model/dpns.rs`:
- [BLOCKING] line 1: UTF-8 panic in strip_dash_suffix and has_dash_suffix (CONVERGENT — NOT FIXED)
  Both functions slice strings at byte offsets (input.len() - DASH_SUFFIX.len()) without checking is_char_boundary(). For inputs like "tëstab" (7 bytes), the split point (2) falls inside the 2-byte ë character → panic. Fix commit cb4652e8 claims to address this but only extracted a constant; the byte-offset slicing is identical. DPNS labels are ASCII-only on-chain, but these functions run on raw user input BEFORE platform validation, so any multi-byte character in search/contact fields triggers an unrecoverable crash.

In `src/backend_task/dashpay/contact_info.rs`:
- [SUGGESTION] line 1: rootEncryptionKeyIndex written from authentication signing key, not encryption root key
  Line 435 writes `signing_key.id()` into `rootEncryptionKeyIndex`. The signing_key is an AUTHENTICATION-purpose key (lines 413-424), but rootEncryptionKeyIndex per DIP-15 should reference the identity key used as the root of the encryption key derivation tree. The read path at line 353 ignores this value (`_root_idx`), so self-decryption works, but other DashPay clients (e.g., mobile wallets) that rely on this field to find the correct encryption key will fail to decrypt contactInfo.

In `src/backend_task/dashpay/contact_requests.rs`:
- [SUGGESTION] line 1: Hardcoded DPNS contract ID breaks non-mainnet contact requests
  resolve_username_to_identity() hardcodes the mainnet DPNS contract ID (GWRSAVFMjXx8HpQFaNJMqBV7MBgMK4br5UESsB4S31Ec) and re-fetches it from the network, while every other DPNS lookup uses app_context.dpns_contract (which is network-aware). On testnet/devnet, this hardcoded ID won't match, causing contact request username resolution to fail.
```
</details>

**Review by thepastaclaw — COMMENTED** [review]
## Code Review

Clean fix commit addressing prior review findings. The DASH_SUFFIX constant eliminates the magic number and makes the byte-offset slicing self-documenting. The UTF-8 safety concern is resolved in practice: `.dash` is always ASCII, so `input.len() - DASH_SUFFIX.len()` always lands on a char boundary regardless of multi-byte characters in the label. MIN_PLAINTEXT_SIZE corrected from 17 to 16. Good additional test coverage for the `.dash` edge case. The remaining prior findings (rootEncryptionKeyIndex, hardcoded DPNS contract ID, sentinel padding) are not addressed in this push.

*Reviewed commit: cb4652e8*

**Review by thepastaclaw — COMMENTED** [review]
## Code Review

Excellent follow-up fixing all DPNS-related findings from the prior review. Byte-index slicing replaced with safe `.get()` + `.is_some_and()` pattern — eliminates the UTF-8 panic risk entirely. API asymmetry between `has_dash_suffix` and `strip_dash_suffix` resolved by adding trim to `strip_dash_suffix`. Non-ASCII test case added. Clean merge of v1.0-dev (includes #813 SPV test fix). The remaining prior findings (rootEncryptionKeyIndex, hardcoded DPNS contract ID, sentinel padding) are not in scope for this fix commit.

*Reviewed commit: d7fea390*

File: src/backend_task/dashpay/contact_requests.rs [hunk 2]
```diff
@@ -188,7 +188,8 @@ pub async fn send_contact_request_with_proof(
     qr_auto_accept: Option<AutoAcceptProofData>,
 ) -> Result<BackendTaskSuccessResult, TaskError> {
     // Step 1: Resolve the recipient identity
-    let to_identity = if to_username_or_id.ends_with(".dash") {
+    let to_username_or_id = to_username_or_id.trim().to_string();
+    let to_identity = if crate::model::dpns::has_dash_suffix(&to_username_or_id) {
         // It's a complete username, resolve via DPNS
         resolve_username_to_identity(sdk, &to_username_or_id).await?
     } else {
```
[the remark this answers is not in the document]
Comment (lklimek) [reply src/backend_task/dashpay/contact_requests.rs:202]: Fixed in 83a5c2c3 — inputs containing a `.` without a `.dash` suffix now return `DashPayError::InvalidUsername` early, avoiding unnecessary DPNS queries.

<sub>🤖 Co-authored by [Claudius the Magnificent](https://github.com/lklimek/claudius) AI Agent</sub>

File: src/backend_task/dashpay/contact_info.rs [hunk 3]
```diff
@@ -387,7 +408,8 @@ pub async fn create_or_update_contact_info(
     let encrypted_private_data = encrypt_private_data(&private_data.serialize(), &private_data_key)
         .map_err(|e| TaskError::EncryptionError { detail: e })?;
 
-    // Get signing key
+    // Get signing key — accept any key type (BLS, ECDSA, EDDSA) since
```
[the remark this answers is not in the document]
Comment (lklimek) [reply src/backend_task/dashpay/contact_info.rs:444]: Fixed in 83a5c2c3 — now calls `validate_contact_info_field_sizes()` after encryption and before broadcasting. Oversized data returns `DashPayError::ContactInfoValidationFailed` with a user-friendly message.

<sub>🤖 Co-authored by [Claudius the Magnificent](https://github.com/lklimek/claudius) AI Agent</sub>

**Review by thepastaclaw — CHANGES_REQUESTED** [review]
## Code Review

I found one new blocking issue in the current SHA: profile search still resolves DPNS hits through `owner_id()` even though this PR correctly switches direct username resolution to `records.identity`, so transferred names can still surface the wrong profile in search results. I also validated the live CodeRabbit thread: the raw `u8` length/count overflow in contact-info serialization and the missing legacy error-helper wiring are both real, while the `InvalidDocument { reason }` complaint is not — the formatted `reason` is internal detail on a variant whose user-facing text already comes from the enum-level display string.

I validated existing CodeRabbit comments separately in-thread instead of restating them here.

*Reviewed commit: 142a0422*

🔴 1 blocking

### 1 additional finding

<details>
<summary>🔴 <b>blocking:</b> Profile search still resolves transferred names through `owner_id()`</summary>

`src/backend_task/dashpay/profile.rs (lines 486-490)`

`search_profiles()` now normalizes the input correctly, but it still derives the profile identity from `document.owner_id()` when turning a DPNS domain result into a user profile. This PR explicitly fixes the same ownership mismatch elsewhere by switching direct username resolution to `records.identity`, because transferred DPNS names can diverge from the original document owner. Leaving `search_profiles()` on `owner_id()` means a transferred username can still return the previous owner's profile from search even after add-contact/name lookup was fixed.

<details>
<summary>💡 Suggested change</summary>

```suggestion
Use the DPNS document's `records.identity` field here too (ideally through a shared helper), and only fall back to `owner_id()` if the field is absent.
```

</details>

</details>

<details>
<summary>🤖 Prompt for all review comments with AI agents</summary>

```
These findings are from an automated code review. Verify each finding against the current code and only fix it if needed.

In `src/backend_task/dashpay/profile.rs`:
- [BLOCKING] lines 486-490: Profile search still resolves transferred names through `owner_id()`
  `search_profiles()` now normalizes the input correctly, but it still derives the profile identity from `document.owner_id()` when turning a DPNS domain result into a user profile. This PR explicitly fixes the same ownership mismatch elsewhere by switching direct username resolution to `records.identity`, because transferred DPNS names can diverge from the original document owner. Leaving `search_profiles()` on `owner_id()` means a transferred username can still return the previous owner's profile from search even after add-contact/name lookup was fixed.
```
</details>

File: src/backend_task/dashpay/contact_requests.rs [hunk 4]
```diff
@@ -517,57 +524,74 @@ pub async fn send_contact_request_with_proof(
...
-    // Get the identity ID from the DPNS document
-    let identity_id = document.owner_id();
+        .ok_or_else(|| {
+            TaskError::DashPay(DashPayError::InvalidDocument {
+                reason: format!(
+                    "DPNS document for '{}' is missing records.identity field",
+                    username
+                ),
+            })
+        })?;
```
[the remark this answers is not in the document]
Comment (thepastaclaw) [reply src/backend_task/dashpay/contact_requests.rs:599]: I dont think this is user-facing in the current code path. `DashPayError::InvalidDocument { reason }` still renders through the enums fixed display string, and the formatted `reason` is carried as internal detail rather than shown directly to the user, so I dont see a user-visible regression from keeping the `records.identity` note in that payload here.

**Review by thepastaclaw — CHANGES_REQUESTED** [review]
## Code Review

I validated the current SHA and the main correctness issue from the prior review is still present: `search_profiles()` still resolves DPNS hits through `owner_id()` and `normalizedLabel`, so transferred names can still return and display the wrong profile. The contact-info serialization overflow and the missing `ContactInfoValidationFailed` helper wiring also remain real in the current code, and both match live CodeRabbit comments. I did not keep the multi-label `.dash` validation concern because it is lower-confidence and not needed for the review outcome.

I validated existing CodeRabbit comments separately in-thread instead of restating them here.

*Reviewed commit: 5cc50460*

🔴 1 blocking

### 1 additional finding

<details>
<summary>🔴 <b>blocking:</b> Profile search still resolves transferred DPNS names through `owner_id()`</summary>

`src/backend_task/dashpay/profile.rs (lines 486-493)`

`search_profiles()` still derives its result identity from `document.owner_id()` and its display name from `normalizedLabel`. For transferred DPNS names, the authoritative current identity lives in `records.identity`, not the document owner, and the human-facing label should come from the original `label` field rather than the normalized homograph-safe one. The rest of this PR already switched direct username/contact resolution to `records.identity`, so profile search still disagrees with the corrected behavior and can show the wrong person's profile for a transferred name.

<details>
<summary>💡 Suggested change</summary>

```suggestion
Use the DPNS document's `records.identity` field here too (ideally through a shared helper), and render the user-facing name from `label` rather than `normalizedLabel`.
```

</details>

</details>

<details>
<summary>🤖 Prompt for all review comments with AI agents</summary>

```
These findings are from an automated code review. Verify each finding against the current code and only fix it if needed.

In `src/backend_task/dashpay/profile.rs`:
- [BLOCKING] lines 486-493: Profile search still resolves transferred DPNS names through `owner_id()`
  `search_profiles()` still derives its result identity from `document.owner_id()` and its display name from `normalizedLabel`. For transferred DPNS names, the authoritative current identity lives in `records.identity`, not the document owner, and the human-facing label should come from the original `label` field rather than the normalized homograph-safe one. The rest of this PR already switched direct username/contact resolution to `records.identity`, so profile search still disagrees with the corrected behavior and can show the wrong person's profile for a transferred name.
```
</details>

File: src/backend_task/dashpay/contact_requests.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (thepastaclaw) [reply src/backend_task/dashpay/contact_requests.rs:599]: I don't see a current-SHA user-facing issue here. The `records.identity` text is only stored in the internal `InvalidDocument.reason` payload; the active user-facing path still renders through `DashPayError::InvalidDocument`'s generic display text, so this field name is not exposed to users.

[the remark this answers is not in the document]
Comment (thepastaclaw) [reply src/backend_task/dashpay/contact_requests.rs:599]: No further action needed here. This follow-up already retracts the earlier concern by noting that `InvalidDocument.reason` is diagnostic-only and not shown to users on the current SHA.

[the remark this answers is not in the document]
Comment (thepastaclaw) [reply src/backend_task/dashpay/contact_requests.rs:599]: I don't see a current-SHA user-facing issue here. The `records.identity` text is only stored in the internal `InvalidDocument.reason` payload; the active user-facing path still renders through `DashPayError::InvalidDocument`'s generic display text, so this field name is not exposed to users.

File: src/backend_task/dashpay/errors.rs [hunk 5]
```diff
@@ -170,6 +170,10 @@ pub enum DashPayError {
     /// A contact request has already been sent to this recipient.
     #[error("You have already sent a contact request to '{to}'. Please wait for them to respond.")]
     ContactRequestAlreadySent { to: String },
+
+    /// Encrypted contact info fields exceed DashPay contract limits.
+    #[error("Contact info is too large to save. Try shortening your nickname or note.")]
+    ContactInfoValidationFailed { errors: Vec<String> },
```
[the remark this answers is not in the document]
Comment (thepastaclaw) [reply src/backend_task/dashpay/errors.rs:179]: Valid, and on the current SHA it now applies to both `ContactInfoValidationFailed` and the newly added `CannotContactSelf` variant: both still fall through `user_message()` / `requires_user_action()` even though the UI continues to call those helpers.

File: src/backend_task/dashpay/contact_requests.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (thepastaclaw) [reply src/backend_task/dashpay/contact_requests.rs:599]: No further action needed here. This follow-up already retracts the earlier concern by noting that `InvalidDocument.reason` is diagnostic-only and not shown to users on the current SHA.

File: src/backend_task/dashpay/contact_info.rs [hunk 6]
```diff
@@ -387,7 +408,20 @@ pub async fn create_or_update_contact_info(
...
+        &encrypted_user_id,
+        &encrypted_private_data,
+    );
+    if !validation.is_valid {
+        return Err(TaskError::DashPay(
+            DashPayError::ContactInfoValidationFailed {
+                errors: validation.errors,
+            },
+        ));
+    }
```
[the remark this answers is not in the document]
Comment (lklimek) [reply src/backend_task/dashpay/contact_info.rs:442]: Fixed in cd90c212 — `serialize()` now returns `Result<Vec<u8>, DashPayError>` and validates all three lengths (nickname, note, accounts) fit in `u8` before casting. Overflow returns `ContactInfoValidationFailed` with a descriptive message.

<sub>🤖 Co-authored by [Claudius the Magnificent](https://github.com/lklimek/claudius) AI Agent</sub>

File: src/backend_task/dashpay/errors.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (lklimek) [reply src/backend_task/dashpay/errors.rs:179]: Fixed in cd90c212 — both `ContactInfoValidationFailed` and `CannotContactSelf` are now wired into `user_message()` and `requires_user_action()`.

<sub>🤖 Co-authored by [Claudius the Magnificent](https://github.com/lklimek/claudius) AI Agent</sub>

**Review by thepastaclaw — CHANGES_REQUESTED** [review]
## Code Review

The follow-up fixes from the prior review are present, but one DPNS search path still uses the wrong fields when turning domain documents into user-facing search results. `search_profiles()` remains inconsistent with the transfer-aware resolution logic added elsewhere in this SHA, so the PR still has one correctness issue that should block approval.

*Reviewed commit: cd90c212*

🔴 1 blocking

### 1 additional finding

<details>
<summary>🔴 <b>blocking:</b> Profile search still resolves DPNS results through `owner_id()` and `normalizedLabel`</summary>

`src/backend_task/dashpay/profile.rs (lines 486-493)`

`search_profiles()` still converts each DPNS domain document into a result by reading `document.owner_id()` and `normalizedLabel`. In this same SHA, the other DPNS resolution paths explicitly switched to `records.identity` because that is the authoritative current identity after a name transfer, and they use `label` for display because `normalizedLabel` is only the homograph-safe lookup key. Leaving this path on the old fields means a transferred name can still load the previous owner's profile, and even non-transferred names can be rendered as the normalized slug instead of the user-facing username.

</details>

<details>
<summary>🤖 Prompt for all review comments with AI agents</summary>

```
These findings are from an automated code review. Verify each finding against the current code and only fix it if needed.

In `src/backend_task/dashpay/profile.rs`:
- [BLOCKING] lines 486-493: Profile search still resolves DPNS results through `owner_id()` and `normalizedLabel`
  `search_profiles()` still converts each DPNS domain document into a result by reading `document.owner_id()` and `normalizedLabel`. In this same SHA, the other DPNS resolution paths explicitly switched to `records.identity` because that is the authoritative current identity after a name transfer, and they use `label` for display because `normalizedLabel` is only the homograph-safe lookup key. Leaving this path on the old fields means a transferred name can still load the previous owner's profile, and even non-transferred names can be rendered as the normalized slug instead of the user-facing username.
```
</details>
